### ISSUE TITLE
Large FixedSizeVec overflows stack

### ISSUE THREAD

**Issue by jasonashton** [eclipse-iceoryx/iceoryx2#220]
## Required information

**Operating system:**
* MacOS Sonoma 14.5  
`Darwin Macbook-Pro.local 23.5.0 Darwin Kernel Version 23.5.0: Wed May  1 20:12:58 PDT 2024; root:xnu-10063.121.3~5/RELEASE_ARM64_T6000 arm64`

**Rust version:**
`rustc 1.78.0 (9b00956e5 2024-04-29)`

**Cargo version:**
`cargo 1.78.0 (54d8815d0 2024-03-26)`

**iceoryx2 version:**
`main`

**Observed result or behaviour:**
When creating a large (1mb or greater) FixedSizeVec, the stack overflows

**Expected result or behaviour:**
This is more of a question, but how can we send larger data over the IPC? I'm looking to transfer images between processes which needs much larger data sizes, but that overflows the stack. The answer is usually to use a Vec on the heap but that doesn't seem to be an option for IPC.

**Conditions where it occurred / Performed steps:**
In `iceoryx2-bb/container/tests/vec_tests.rs` set 
```bash
// const SUT_CAPACITY: usize = 128;
const SUT_CAPACITY: usize = 6220800;
type Sut = FixedSizeVec<usize, SUT_CAPACITY>;
```
and run `cargo test --test vec_tests`:

```bash
thread 'fixed_size_vec_capacity_is_correct' has overflowed its stack
fatal runtime error: stack overflow
error: test failed, to rerun pass `-p iceoryx2-bb-container --test vec_tests`
```

**Comment by elfenpiff** [comment]
@jasonashton Thanks a lot for pointing out this issue. I had it once on the radar but lost track of it.

What we in the end need is something like a real placement new, but I think the RFC to it was dismissed in 2018. So, we have to define something that works with our shared memory use case.

In the meantime, you can use the slice API where you shouldn't encounter this problem, see the example here: https://github.com/eclipse-iceoryx/iceoryx2/tree/main/examples/rust/publish_subscribe_dynamic_data

**Comment by elfenpiff** [comment]
@elBoberido mentioned to me that this approach should work in release mode:

```rust
let service = zero_copy::Service::new(&service_name)
    .publish_subscribe()
    .open_or_create::<FixedSizeVec<u64, 100000000>>()?;
// ...
let sample = unsafe {
    sample
        .payload_mut()
        .as_mut_ptr()
        .write(FixedSizeVec::default());

    sample.assume_init()
};
```

It is restricted to release mode. Here the compiler optimizes the copy away and implicitly writes directly into the shared memory.

**Comment by jasonashton** [comment]
@elfenpiff thanks for the quick response! I tried the second example, it works for me as a `bin` but not a test (even with `--release`, any ideas why?

This works for me, it sends and receives
```bash
use core::time::Duration;
use iceoryx2::prelude::*;
use iceoryx2_bb_container::vec::FixedSizeVec;

const CYCLE_TIME: Duration = Duration::from_secs(1);
const COMMON_IMAGE_SIZE: usize = 3 * 1080 * 1920;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let service_name = ServiceName::new("my/big/topic4")?;

    let service = zero_copy::Service::new(&service_name)
        .publish_subscribe::<FixedSizeVec<u8, COMMON_IMAGE_SIZE>,>()
        .open_or_create()?;

    let publisher = service.publisher().create()?;

    let mut counter: u64 = 0;

    while let Iox2Event::Tick = Iox2::wait(CYCLE_TIME) {
        counter += 1;
        let mut sample = publisher.loan_uninit()?;
        let mut img = FixedSizeVec::<u8, COMMON_IMAGE_SIZE>::default();
        for i in 0..img.capacity() {
            img.push((i % 255) as u8);
        }
        let sample = unsafe {
            sample
                .payload_mut()
                .as_mut_ptr()
                .write(img);
            sample.assume_init()
        };
        sample.send()?;
        println!("Sent sample {} ...", counter);
    }
    println!("exit");
    Ok(())
}

```

But this simple test fails with a stack overflow:
```bash
    #[test]
    fn test_big_fixed_size_vec() {
        const COMMON_IMAGE_SIZE: usize = 3 * 1080 * 1920;
        let mut img = FixedSizeVec::<u8, COMMON_IMAGE_SIZE>::default();
        for i in 0..img.capacity() {
            img.push((i % 255) as u8);
        }
        assert_eq!(img.len(), img.capacity());
    }
```

On the topic of the dynamic pub/sub example, I don't fully understand how it works, it seems like it is essentially just breaking up the larger data into multiple publishes? So based on my understanding I'd have to reassemble the data on the other end, is that correct?

Lastly, I believe we used iceoryx zero-copy RMW for ROS2 (RouDI) (though on linux, not mac) at a previous company I was at, and we were sending multiple large images in a single message through zero-copy, and I had never run into an issue like this. Is there some difference here? Is it a iceoryx2, or mac, or rust specific issue?

Thanks for the help, and for the great project!

**Comment by jasonashton** [comment]
Just another data point, I tried the dynamic publisher but just set max slice length to the size of the vec and that seems to work for me:
```bash
use core::time::Duration;
use iceoryx2::prelude::*;

const CYCLE_TIME: Duration = Duration::from_secs(1);

const COMMON_IMAGE_SIZE : usize = 3 * 1920 * 1080;

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let service_name = ServiceName::new("/big/dynamic/service")?;

    let service = zero_copy::Service::new(&service_name)
        .publish_subscribe::<[u8]>()
        .open_or_create()?;

    let publisher = service
        .publisher()
        .max_slice_len(COMMON_IMAGE_SIZE)
        .create()?;

    let mut counter = 1;
    let mut vec = Vec::<u8>::with_capacity(COMMON_IMAGE_SIZE);
    for i in 0..vec.capacity() {
        vec.push((i % 255) as u8);
    }

    while let Iox2Event::Tick = Iox2::wait(CYCLE_TIME) {
        counter += 1;

        let sample = publisher.loan_slice_uninit(COMMON_IMAGE_SIZE)?;
        let sample = sample.write_from_fn(|byte_idx| vec[byte_idx]);
        sample.send()?;
        println!("Send sample {} ...", counter);
    }

    println!("exit");
    Ok(())
}

```
The dynamic subscriber example prints `received 6220800 bytes`

**Comment by elfenpiff** [comment]
@jasonashton 

>  I tried the second example, it works for me as a bin but not a test (even with --release, any ideas why?

I think that `cargo test --release` is slightly different than the `cargo run --release` mode but I cannot name any specific differences.

> On the topic of the dynamic pub/sub example, I don't fully understand how it works, it seems like it is essentially just breaking up the larger data into multiple publishes? 

No, the idea is to send a slice, and you can define the worst-case size on the publisher's side. The example is that you want to send camera data, and at runtime you detect a 720p camera. Therefore, you know your image size is 4 * 1280 * 720 - so you create a publisher with this worst-case size. After some runtime, the camera is disconnected and replaced with a 4k camera, and your image size becomes 4 * 3840 * 2160. You remove the old publisher and create a new one with a larger worst-case size.
If the old camera is connected again you can either stick to the current worst-case size or remove the new publisher again and create a new one with a smaller size.

I think I need to write a blog article to clarify this a bit more in detail.

In the future this shall all happen behind the scenes so that you do not have to recreate a publisher anymore. But for now it is already a big improvement compared to iceoryx1.

> Lastly, I believe we used iceoryx zero-copy RMW for ROS2 (RouDI) (though on linux, not mac) at a previous company I was at, and we were sending multiple large images in a single message through zero-copy, and I had never run into an issue like this. Is there some difference here? Is it a iceoryx2, or mac, or rust specific issue?

It is a rust-specific issue. Take a look at this here: https://github.com/rust-lang/rust/issues/53827
This is exactly the problem behind the scenes since the `FixedSizeVector<u8, 100000000>` has an array member of `[MaybeUninit<u8>, 100000000]` that overflows the stack when you want to put it into the heap like

```rust
let heap_array = Box::new([0; 100000000]);
```

But we already started solving this. The idea is to introduce a `PlacementNew` trait

```rust
pub trait PlacementNew: Sized {
    type NewParam;
    
    fn placement_new(this: &mut MaybeUninit<Self>, param: &Self::NewParam);
}
```

in combination with a proc derive `placement_new` macro so that the user can use it like:

```rust
#[derive(placement_new)]
struct MyTransmissionType {
  data: FixedSizeVec<u64, 1234>,
  whatever: FixedSizeString<4567>,
}
```

> Just another data point, I tried the dynamic publisher but just set max slice length to the size of the vec and that seems to work for me:

Perfect :)

In the future we want to integrate also `RelocatableVec` which you could use like this

```rust
struct CameraData {
  image: RelocatableVec<u8>,
  camera_details: FixedSizeString<128>,
  whatever: u64
}

// ...

let sample = publisher.loan()?;
sample.image.reserve(SIZE_OF_CAMERA_IMAGE);
// ...
sample.image.push(123);
```

**Comment by elBoberido** [comment]
@jasonashton the example with iceoryx uses shared memory and in release builds the compiler is able to optimize the copy from the stack to the shared memory away and writes the vector directly to the shared memory. In the test this is not possible since it will stay on the stack. You could circumvent this by using `Box<FixedSizeVec<u8, COMMON_IMAGE_SIZE>>` for the tests but there would still be the issue with the stack overflow in debug builds. Using `ulimit -s unlimited` would be a workaround but we need a proper solution. Ideally, Rust would provide something similar to the C++ placement new.

**Comment by jasonashton** [comment]
> No, the idea is to send a slice, and you can define the worst-case size on the publisher's side. The example is that you want to send camera data, and at runtime you detect a 720p camera. Therefore, you know your image size is 4 * 1280 * 720 - so you create a publisher with this worst-case size. After some runtime, the camera is disconnected and replaced with a 4k camera, and your image size becomes 4 * 3840 * 2160. You remove the old publisher and create a new one with a larger worst-case size.

This is what I ended up implementing, I abstracted the publisher to automatically do this. The benefit of this is actually that I can publish non-fixed sizes like a serialized protobuf, which actually works slightly better for us, though I'm sure at the cost of some performance?

> in combination with a proc derive placement_new macro so that the user can use it like:

This is close to what I was originally trying to implement (though with the FixedSizeVec) as I'm used to the requirement of compile time fixed sizes from Iceoryx / RouDI. The relocatable vec solution seems good too.

Thanks for the help! I'm currently unblocked now, but looking forward to trying out the fixes when they land.

**Comment by elBoberido** [comment]
Oh, you already used iceoryx with RouDi? Can you share some details about your project?

We do not have numbers for the impact of creating new publishers. In theory it should not be too expensive but maybe some optimizations are necessary.

**Comment by jasonashton** [comment]
It was at a previous company working on ROS2-based robots. We had 8 uncompressed images we needed to get from one node to another and zero copy was necessary as otherwise the software would quickly grind to a halt using DDS. I wasn't the engineer directly setting it up but I was working around it. What I have working now with Iceoryx2 in a week is farther and more flexible than what took months of work and troubleshooting previously.

**Comment by elBoberido** [comment]
Cool. Would be great to get something like this as success/user story on our website ... and with our website I mean ekxide.io :)

**Comment by elfenpiff** [comment]
@jasonashton We implemented a `PlacementDefault` trait and derive proc-macro to address the issue. It is implemented for all iceoryx2 containers, and a working example is here: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/examples/rust/complex_data_types/complex_data_types.rs

I also added a FAQ entry: https://github.com/eclipse-iceoryx/iceoryx2/blob/main/FAQ.md#my-transmission-type-is-too-large-encounter-stack-overflow-on-initialization

For now, I would define this as the iceoryx2 idiomatic approach to initialize large structs in shared memory.

But as @elBoberido already found out, in release mode, the compiler optimizes this so that the stack overflow should not pose a problem.

If you have time to play around with it, I would be happy to receive some feedback.

**Comment by berkgercek** [comment]
Hi there @elfenpiff , I am currently seeing the same behavior while using `iceoryx2 = "0.6.1"` (`rustc 1.89.0 (29483883e 2025-08-04)`, and `cargo 1.89.0 (c24e10642 2025-06-23)`, while using Fedora 42 `Linux 6.16.3-200.fc42.x86_64 x86_64` ), and using an e.g. 1920x1080 RGBA `FixedSizeVec` causes the stack to overflow:

```
thread 'main' has overflowed its stack
fatal runtime error: stack overflow, aborting
Aborted (core dumped)
```

Additionally when running the CodeLLDB debugger on the binary I get this error:
`signal SIGSEGV: address not mapped to object (fault address=0x7fffff7feea8)`

I'm in a similar use case to @jasonashton , and am trying to create an IPC system to transfer vertices/indices/textures to a rendering process, which will then render them to the display (necessary because I need the main thread and MacOS/iOS require windows to be launched from the main thread). 

It seems, at least in my case, that PR didn't fully solve the problem.

I was working on a POC example for this layer, and this is the full code base:

```rust
use std::time::{Duration, Instant};

use iceoryx2::port::publisher::Publisher;
use iceoryx2::port::subscriber::Subscriber;
use iceoryx2::prelude::*;
use iceoryx2_bb_container::vec::FixedSizeVec;

pub const MAX_TEXTURE_SIZE: usize = 1920 * 1080 * 4; // 4K Assuming 4 bytes per pixel (RGBA)
pub const MAX_VERTICES: usize = 100; // Maximum number of vertices
pub const MAX_INDICES: usize = 100; // Maximum number of indices
pub const MAX_DISPLAYOBJECT_SIZE: usize = MAX_VERTICES + MAX_INDICES + MAX_TEXTURE_SIZE + 2;

#[derive(Debug, Clone, Default, PlacementDefault, ZeroCopySend)]
#[repr(C)]
pub struct DisplayObject {
    pub command_id: u16,
    pub vertices: FixedSizeVec<f32, MAX_VERTICES>,
    pub indices: FixedSizeVec<u32, MAX_INDICES>,
    pub texture: FixedSizeVec<u8, MAX_TEXTURE_SIZE>,
}

pub struct IpcClient {
    pub node: Node<ipc::Service>,
    pub command_publisher: Publisher<ipc::Service, DisplayObject, ()>,
}

impl IpcClient {
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        let node = NodeBuilder::new().create()?;
        let command_service = node
            .service_builder(&"GeminiDisplayService".try_into()?)
            .publish_subscribe::<DisplayObject>()
            .max_publishers(1)
            .max_subscribers(16)
            .open_or_create()?;
        let command_publisher = command_service
            .publisher_builder()
            // .initial_max_slice_len(MAX_TEXTURE_SIZE)
            // .allocation_strategy(AllocationStrategy::PowerOfTwo)
            // .unable_to_deliver_strategy(UnableToDeliverStrategy::Block)
            .create()?;
        Ok(IpcClient {
            node,
            command_publisher,
        })
    }

    pub fn send_command_complex(
        &self,
        command_id: u16,
        vertices: FixedSizeVec<f32, MAX_VERTICES>,
        indices: FixedSizeVec<u32, MAX_VERTICES>,
        texture: FixedSizeVec<u8, MAX_TEXTURE_SIZE>,
    ) -> Result<(), Box<dyn std::error::Error>> {
        let mut sample = self.command_publisher.loan_uninit()?;
        // Copy serialized data into the slice
        unsafe { DisplayObject::placement_default(sample.payload_mut().as_mut_ptr()) };
        let mut sample = unsafe { sample.assume_init() };
        let payload = sample.payload_mut();
        payload.command_id = command_id;
        payload.indices = indices;
        payload.vertices = vertices;
        payload.texture = texture;

        // Send the sample
        sample.send()?;
        Ok(())
    }
}

pub struct IpcServer {
    pub node: Node<ipc::Service>,
    pub command_subscriber: Subscriber<ipc::Service, DisplayObject, ()>,
}

impl IpcServer {
    pub fn new() -> Result<Self, Box<dyn std::error::Error>> {
        let node = NodeBuilder::new().create()?;

        // Create command subscriber for dynamic byte slices
        let command_service = node
            .service_builder(&ServiceName::new("GeminiDisplayService")?)
            .publish_subscribe::<DisplayObject>()
            .max_publishers(1)
            .max_subscribers(16)
            .open_or_create()?;

        let command_subscriber = command_service.subscriber_builder().create()?;

        Ok(Self {
            node,
            command_subscriber,
        })
    }

    pub fn poll_command(&self) -> Result<DisplayObject, Box<dyn std::error::Error>> {
        let start_t = Instant::now();
        while self.node.wait(Duration::from_millis(10)).is_ok() {
            if let Ok(Some(disp_obj)) = self.command_subscriber.receive() {
                return Ok(disp_obj.payload().clone());
            }
            if start_t.elapsed() >= Duration::from_secs(5) {
                return Err("Timed out waiting for command".into());
            }
        }
        return Err("Timed out waiting for command".into());
    }
}
```

The binary simply creates a client and server, then attempts to send a `DisplayObject` using the `send_command_complex` method, then confirms it was received. The overflow happens immediately at the start of the binary however.

**Comment by elBoberido** [comment]
@berkgercek do you have a full example. I think the issue is not related to iceoryx2 itself but to your `send_command_complex` function. It takes the `texture` by value and copies it to the iceoryx2 sample. This means it is already on the stack before it reaches the iceoryx2 sample.

In order to achieve true zero-copy, you need to do the `loan` at the place where you generate the texture. Since you are using vectors and an u32, you can use the simplified `loan` API instead of the `loan_uninit`. This will also take care of the default initialization of your data and prevent the stack overflow issue.

You then have to generate the data directly into the `texture` field of the sample.

**Comment by berkgercek** [comment]
The Texture/Indices/Vertices are all being created in a simple test script in the `main` function:

```rust
use crate::{MAX_INDICES, MAX_TEXTURE_SIZE, MAX_VERTICES};
use oryx_testing::*;

pub fn main() {
    println!("Starting IPC test");
    let server = IpcServer::new().unwrap();
    println!("Server started with node {:?}", server.node);

    let client = IpcClient::new().unwrap();
    println!("Client started and connected");

    let command_id = 1;
    let vertices = {
        let mut v = FixedSizeVec::<f32, MAX_VERTICES>::new();
        v.fill(0.0f32);
        v
    };
    let indices = {
        let mut v = FixedSizeVec::<u32, MAX_INDICES>::new();
        v.fill(1u32);
        v
    };
    let texture = {
        let mut v = FixedSizeVec::<u8, MAX_TEXTURE_SIZE>::new();
        v.fill(128u8);
        v
    };
    let ref_id = command_id.clone();

    // Test sending a command
    client
        .send_command_complex(command_id, vertices, indices, texture)
        .unwrap();
    println!("Object command sent");

    // Test polling the command on the server
    let received_command = server.poll_command().unwrap();
    println!("Show command received");

    assert_eq!(received_command.command_id, ref_id);
}
```
I've tried this example with a direct creation of the `DisplayObject` as well (the `send_command` method took the object as it's only argument, and simply sent it) but the same stack overflow occurred. Part of why I removed the allocation of the object was to figure out whether the issue was happening at the instantiation of the struct.

Edited: Accidentally pasted the current version using the dynamic data size interface

**Comment by berkgercek** [comment]
@elBoberido well it seems you were right; because I wasn't using the default placement it was indeed the allocation of the initial `FixedSizeVec` causing the stack to overflow. When I switched the above script to use `vec![0.0u8; MAX_TEXTURE_SIZE]` for the initial argument to the function the code ran as expected. I see what you're saying regarding generating the texture directly into the field, and going forward I'll do that instead.

Why is it that the initial `FixedSizeVec` assignment causes an overflow? Is it because the `PlacementDefault` trait isn't being used?

Thank you! I'm very new to Rust and in general lower level programming, so much of this is quite alien to me.

**Comment by elBoberido** [comment]
@berkgercek yeah, it is also required to get rid of the copy into the sample.

You basically need to do something like this

```rs
let client = IpcClient::new().unwrap();
println!("Client started and connected");

let sample = client.loan().unwrap();
sample.texture.fill(128u);

sample.send();
```

If you do it right now, you will always copy the 16MB into the sample and lose performance due to the copy.

**Comment by elBoberido** [comment]
Oh, btw. the `PlacementDefault` only helps if the data structure is not placed on the stack. So it works only if it is on the heap or in the case of iceoryx2 in the shared memory.

**Comment by berkgercek** [comment]
While I get that the `publisher.loan` method would be ideal, unfortunately when I use that method I immediately run in to the stack overflowing again:

```rust
    pub fn send_command_complex(&self, command_id: u16) -> Result<(), Box<dyn std::error::Error>> {
        let start_t = Instant::now();
        let mut sample = self.command_publisher.loan()?;
        let elapsed = start_t.elapsed();
        println!("Allocating the sample took {elapsed:?}");
        sample.command_id = command_id;
        sample.indices.fill(128u32);
        sample.vertices.fill(0.5f32);
        sample.texture.fill(128u8);
        let elapsed = start_t.elapsed();
        println!("Filling the sample with {MAX_DISPLAYOBJECT_SIZE} bytes took {elapsed:?}");
        // Send the sample
        sample.send()?;
```

This causes an overflow immediately as compared to the placement default method. Ditto if I use `sample.payload()` as the target for the fill command instead of the sample itself.

In a related note the `.fill(value)` method takes around 4-5ms to actually fill the field with the value. This seems exceptionally slow?

**Comment by elBoberido** [comment]
@berkgercek okay, maybe I misunderstood something with the `placement_default`. If it works for you with `loan_uninit` then you go with it.

Regarding the time for `fill`. Is the measurement done with debug builds?

**Comment by berkgercek** [comment]
No, the initial time using debug builds was 100+ms. The 4-5ms is with a release build, and I measured using the `std::time` module (which I assume is accurate enough at this time scale?)

**Comment by elBoberido** [comment]
Okay, that's indeed quite a lot. We need to investigate this. Can you create a separate issue with a minimal code example?

**Comment by berkgercek** [comment]
I've gone ahead and [created a new issue with a minimal reproducible example](https://github.com/eclipse-iceoryx/iceoryx2/issues/1006). Thanks for the fast responses! It's very encouraging, iceoryx2 fits my use case very well and it's been helpful.

**Comment by elBoberido** [comment]
@berkgercek thanks. Are you allowed to share some details about your use case?

**Comment by berkgercek** [comment]
@elBoberido While I don't think I can share here, I've sent Ekxide an email as we could possibly use support in our development!
